### PR TITLE
overlord: implement Overlord.Settle, chiefly for tests

### PR DESCRIPTION
**Pull request by pedronis**
Overlord.Settle runs first a state engine Ensure and then wait for activities to settle. That's done by waiting for all managers activities to settle while making sure no immediate further Ensure is scheduled. Chiefly for tests. 

This is chiefly for tests, it makes some simplifying assumptions and doesn't care if the waiting loop may busy loop sometimes. Cannot be used together with Overlord.Loop.

In the process:
* renamed Overlord.Run to Loop
* changed Ensure in StateEngine to bubble an error collecting all the errors for each manager Ensure

### CODE REVIEW

**Comment by pedronis** [comment]
needs #686

File: overlord/overlord.go [hunk 1]
```diff
@@ -117,12 +118,31 @@ func (o *Overlord) ensureTimerSetup() {
 	o.ensureNext = time.Now().Add(ensureInterval)
 }
 
-func (o *Overlord) ensureTimerReset() {
+func (o *Overlord) syncEnsurePrep() time.Time {
```
Comment (niemeyer) [inline overlord/overlord.go]: The original method name feels much better. syncEnsurePrep gives no hints about what it actually does.

File: overlord/overlord.go [hunk 2]
```diff
@@ -117,12 +118,31 @@ func (o *Overlord) ensureTimerSetup() {
...
 	o.ensureLock.Lock()
 	defer o.ensureLock.Unlock()
 	now := time.Now()
 	o.ensureTimer.Reset(ensureInterval)
 	o.ensureNext = now.Add(ensureInterval)
+	o.ensureRunning++
```
Comment (niemeyer) [inline overlord/overlord.go]: It's not running.. haven't read it all, but this probably has a different meaning.

**Comment by niemeyer** [comment]
This still feels super involved for what seems to be conceptually simple.

Can't we start with something as simple as:
```
func (*Overlord) Settle() {
        done := false
        for !done {
                o.ensureTimerReset()
                stateEngine.Wait()
                o.ensureLock.Lock()
                done = o.ensureNext.After(now)
                o.ensureNext.Unlock()
        }
}
```

?

**Comment by pedronis** [comment]
@niemeyer I don't understand where the first Ensure goes in that code, also that condition afaiu

done = o.ensureNext.After(now)

is naively true

ensureTimerReset just moved ensureNext quite a bit in the future, and ensureBefore may also do the same anyway or not have happened, or there can be another Ensure running parallel to us starting stuff

**Comment by niemeyer** [comment]
Sorry, edited the code to include the missing Ensure in there.

About the remaining comments, I don't quite understand the point. Let's catch up online.

**Comment by pedronis** [comment]
return type ensureError struct { errs []error } from Settle, starting down in the StateEngine and also have the StateEngine log them

File: overlord/overlord.go [hunk 3]
```diff
@@ -167,6 +168,36 @@ func (o *Overlord) Stop() error {
...
 }
 
+// Settle runs first a state engine Ensure and then wait for activities to settle.
+// That's done by waiting for all managers activities to settle while
+// making sure no immediate further Ensure is scheduled. Chiefly for tests.
+// Cannot be used in conjuction with Run.
+func (o *Overlord) Settle() {
+	o.ensureLock.Lock()
+	if o.ensureTimer != nil {
+		panic("cannot use Settle with an ensure loop (Overlord.Run)")
```
Comment (niemeyer) [inline overlord/overlord.go]: `"cannot use Settle concurrently with other Settle or Loop calls"`

Let's rename Run to Loop please.. Run doesn't make clear if it's once or multiple times (Settle also "runs").

File: overlord/overlord.go [hunk 4]
```diff
@@ -167,6 +168,36 @@ func (o *Overlord) Stop() error {
...
+	}
+	o.ensureTimer = time.NewTimer(0)
+	defer func() {
+		o.ensureLock.Lock()
+		defer o.ensureLock.Unlock()
+		o.ensureTimer.Stop()
+		o.ensureTimer = nil
+	}()
+	o.ensureLock.Unlock()
```
Comment (niemeyer) [inline overlord/overlord.go]: For correctness, we should probably avoid the panic inside a mutex region doing something like
```
lock
hadTimer := o.ensureTimer != nil
...
unlock
defer
if hadTimer { panic }
```

**Comment by niemeyer** [comment]
Couple of trivials, and LGTM!

File: overlord/overlord.go [hunk 5]
```diff
@@ -167,6 +165,50 @@ func (o *Overlord) Stop() error {
 	return err1
 }
 
+// Settle runs first a state engine Ensure and then wait for activities to settle.
+// That's done by waiting for all managers activities to settle while
+// making sure no immediate further Ensure is scheduled. Chiefly for tests.
+// Cannot be used in conjuction with Loop.
+func (o *Overlord) Settle() error {
+	func() {
```
Comment (mvo5) [inline overlord/overlord.go]: Just curious: I assume this anonymous function is used for the nice  `Lock(); defer Unlock() ` - or is there another reason?

**Comment by mvo5** [comment]
Looks good.

File: overlord/overlord.go (hunk 5 above)
Comment (pedronis) [inline overlord/overlord.go]: @mvo5 yes, it's nice because is symmetric to the one in the defer

Comment (niemeyer) [inline overlord/overlord.go]: Yes, I'd personally have done without given how simple what's inside it looks, but didn't find it offensive either.

**Pull request merged**